File: ScaffoldingBE-FE/BE/parcial/src/main/java/be/parcial/services/TurnoService.java
package be.parcial.services;

import be.parcial.dtos.ReservaTurnoRequestDTO;
import be.parcial.dtos.TurnoResponseDTO;

import java.util.List;

public interface TurnoService {
    TurnoResponseDTO reservar(String clienteUsername, ReservaTurnoRequestDTO request);
    List<TurnoResponseDTO> misTurnos(String clienteUsername);
    List<TurnoResponseDTO> agenda(String peluqueroUsername);
    TurnoResponseDTO confirmar(String username, Long turnoId);
    TurnoResponseDTO confirmarPorToken(String token);
    TurnoResponseDTO cancelar(String username, Long turnoId);
    TurnoResponseDTO completar(String username, Long turnoId);
}

File: ScaffoldingBE-FE/FE/src/app/shared/services/turno.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Disponibilidad, ReservaTurnoRequest, Turno } from '../models/turno.model';

@Injectable({ providedIn: 'root' })
export class TurnoService {
  private http = inject(HttpClient);
  private readonly base = '/api/turnos';

  disponibilidad(peluqueroId: number, fecha: string, tratamientoId: number): Observable<Disponibilidad> {
    const params = new HttpParams()
      .set('peluqueroId', peluqueroId)
      .set('fecha', fecha)
      .set('tratamientoId', tratamientoId);
    return this.http.get<Disponibilidad>(`${this.base}/disponibilidad`, { params });
  }

  reservar(req: ReservaTurnoRequest): Observable<Turno> {
    return this.http.post<Turno>(this.base, req);
  }

  misTurnos(): Observable<Turno[]> {
    return this.http.get<Turno[]>(`${this.base}/mios`);
  }

  agenda(): Observable<Turno[]> {
    return this.http.get<Turno[]>(`${this.base}/agenda`);
  }

  confirmar(id: number): Observable<Turno> {
    return this.http.patch<Turno>(`${this.base}/${id}/confirmar`, {});
  }

  confirmarPorToken(token: string): Observable<Turno> {
    return this.http.post<Turno>(`${this.base}/confirmar/token/${token}`, {});
  }

  cancelar(id: number): Observable<Turno> {
    return this.http.patch<Turno>(`${this.base}/${id}/cancelar`, {});
  }

  completar(id: number): Observable<Turno> {
    return this.http.patch<Turno>(`${this.base}/${id}/completar`, {});
  }
}
